nifi-nar-bundles/nifi-html-bundle/nifi-html-processors/src/main/java/org/apache/nifi/AbstractHTMLProcessor.java

```diff
@@ -0,0 +1,120 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi;
+
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.io.InputStreamCallback;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class AbstractHTMLProcessor extends AbstractProcessor {
+
+    protected static final String ELEMENT_HTML = "HTML";
+    protected static final String ELEMENT_TEXT = "Text";
+    protected static final String ELEMENT_DATA = "Data";
+    protected static final String ELEMENT_ATTRIBUTE = "Attribute";
+
+    public static final PropertyDescriptor URL = new PropertyDescriptor
+            .Builder().name("URL")
+            .description("Base URL for the HTML page being parsed.")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor CSS_SELECTOR = new PropertyDescriptor
+            .Builder().name("CSS Selector")
+            .description("CSS selector syntax string used to extract the desired HTML element(s).")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(true)
+            .build();
+
+    public static final PropertyDescriptor HTML_CHARSET = new PropertyDescriptor
+            .Builder().name("HTML character encoding")
+            .description("Character encoding of the input HTML")
+            .defaultValue("UTF-8")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .name("original")
+            .description("The original HTML input")
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("Successfully parsed HTML element")
+            .build();
+
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("Failed to parse HTML content")
```

Comment:
I am not sure that I understand the difference between 'failure' and 'invalid html' relationships. I would think that if we "fail to parse HTML content", then it is because "The input HTML syntax is invalid" but those are two separate relationships. Can you expand on the descriptions a bit?